keps/993-two-phase-admission/README.md

```diff
@@ -0,0 +1,246 @@
+
+### Non-Goals
+
+Provide a specific design or implementation for any external controller.
+
+## Proposal
+
+Extend ClusterQueue definition with a list of external controller names that
+need to give green light to admit a workload. Until all of the controllers, which
+identify with these names, don’t give a green light, the workload will not
+be admitted. 
+
+Each workload that goes to such a queue, will get additional conditions, 
+in a dedicated field in the status, each reflecting the go/no-go 
+decision of each of the external controllers. Once
+a controller is confident that the workload can be admitted, it flips the
+status of the corresponding condition from false to true. 
+
+If a controller changes its mind about a workload, it can switch the condition
+back to false and the workload will be immediately suspended.
+
+### User Stories (Optional)
+
+#### Story 1
+
+I want to use the [Cluster Autoscaler ProvisioningRequest API](https://github.com/kubernetes/autoscaler/pull/5848) 
+to ensure that the resources can be provided in full before starting the workload.  
+
+#### Story 2
+
+I want to base workload admission on budget. I have the actual consumption metrics
+in my Prometheus/Datadog/Stackdriver/whatever monitoring system. And weekly
+budget in a CRD, per namespace. I’m happy to modify some small existing sample
+controller, but I don’t want to fork the whole Kueue to adjust some minor details.
+ 
+### Notes/Constraints/Caveats (Optional)
+
+Each of the go/no-go decisions is best-effort and can be reversed at any time. 
+
+### Risks and Mitigations
+
+## Design Details
+
+We will extend ClusterQueueSpec definition by adding a reference to
+AdmissionChecks that the queue needs to perform before admitting a
+workload. The design of AdmissionChecks is similar to IngressClass
+- it allows passing additional parameters, specific to particular checks
+that need to be performed. 
+
+```
+// ClusterQueueSpec defines the desired state of ClusterQueue
+type ClusterQueueSpec struct {
+[...]
+	// List of AdmissionCheck names that needs to be passed
+	// to admit a workload. The order of the checks doesn't matter,
+	// all of them will be started at the same time.
+	AdmissionChecks []string `json:"admissionChecks"`
+}
+
+// WorkloadStatus defines the observed state of Workload
+type WorkloadStatus struct {
+[...]
+	// Status of admission checks, if there are any specified 
+	// for the ClusterQueue in which the workload is queued.
+	AdmissionChecksStatuses []metav1.Condition `json:"admissionChecksStatuses,omitempty"`
+}
+
+
+// Cluster-scoped top-level object defining a check that 
+// can be referenced in ClusterQueue
+type AdmissionCheck struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec AdmissionCheckSpec `json:"spec,omitempty"`
+}
+
+const (
+	Retry metav1.ConditionStatus = “Retry”
+	Reject metav1.ConditionStatus = “Reject”
+	QuotaExhausted metav1.ConditionStatus = "QuotaExhausted"
+)
```

Comment:
nit, move before `AdmissionChecksStatuses`, so that it's more clear that it's about them (IIUC).

But in general, explain what they mean.

Reply:
Explanations added.